src/security/ZeroSensitiveMemory.h

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_SECURITY_ZEROSENSITIVEMEMORY_H
+#define SQUID_SRC_SECURITY_ZEROSENSITIVEMEMORY_H
+
+#include <cstring>
+
+namespace Security {
+
+inline void
```

Comment:
Is there a good reason to inline this function? I do not see one. AFAICT, the volatile pointer trick should prevent inlining of the internal memset() call, so the whole thing is not going to be fully inlined anyway. And once it is not fully inlined, the negative side effects of exposing implementation internals may outweigh the performance improvements from inlining the part before the memset call. In general, I think we want to declare inline only what is likely to be inlined and has tangible performance benefits from inlining; the default is not to inline.

Please discuss or move the function definition to the corresponding .cc file.

N.B. I have added the same change request [earlier](https://github.com/squid-cache/squid/pull/758#discussion_r575687260), but you seemed to ignore it (I am not sure why).

Reply:
I did not ignore, it got lost in the rest of the noise, sorry.

Reply:
Would it be possible to not put the implementation into .cc ? The security part is not a lib thus each potential part of squid would need to made be aware in their respective Makefile.am, complicating things unnecessarily (IMHO).

Reply:
FWIW, this is why I prefer the libcompat approach of using memset_s and implementing the standard behaviour there.

If it is too hard to call the correct policy handler or whatever standard version does that was tricksy. Then using our x prefix and throwing should be safe enough and equivalent to a policy that aborts the processing.

Reply:
> FWIW, this is why I prefer the libcompat approach of using memset_s and implementing the standard behaviour in our replacement.

I think we have already established that it is difficult to correctly implement a standard memset_s replacement while correct use of standard memset_s API requires mainline code changes that are notably more complex than what this PR is currently attempting to do. No such implementation has been offered. I do not understand why all that difficulty and trouble would be _preferred_ to "this" -- a basic question of whether to inline a simple function.

> Then using our x prefix and throwing should be safe enough and equivalent to a policy that aborts the processing.

That is exactly what this PR currently implements, except for the `x` prefix in the name (for obvious reasons). I think it is a much better solution overall, at least for now.

Reply:
> each potential part of squid would need to made be aware in their respective Makefile.am (security/libsecurity.la for instance), complicating things unnecessarily (IMHO).

Yes, every binary that uses libsecurity.la would have to link with it. I agree that more linking is a "complication", of course. However,

1. Inlining alone would not avoid that extra linking -- TextException is defined in `base/` which will have to be linked in if we throw.

2. Adding internal libraries to Squid binaries is not really a "problem". It is inevitability! The only reason we are having this discussion is because helpers like edir_ldapext were relatively recently moved into `src/` (commit 48c5f8e). That move was motivated, in part, by helpers _suffering_ without access to basic Squid-specific functions. We should not start worrying now about adding Squid libraries to Squid helpers (where such addition is warranted). We should enjoy the availability of those libraries.

3. The current inlined solution is not zero-cost either because it unnecessary exposes users to its internal dependencies that some of those users would not have been exposed to otherwise.

4. In its current bare form, ZeroSensitiveMemory() is almost useless -- it provides no protection from the sensitive memory traveling around Squid, being copied, pooled, etc. To properly offer real, tangible protection, we will need to implement non-trivial classes that manage that sensitive memory (and use ZeroSensitiveMemory() internally). Thus, in the long term, some linking will be inevitable even if we restrict ourselves to memory zeroing purposes.

Said that, I suspect that `security/` is actually a wrong place for this new function. That directory/library/namespace cannot possibly contain _everything_ related to Squid security -- that would be too broad and ultimately negate encapsulation benefits. I suspect `security/` should continue to be focused on external communication security while internal security efforts like this new function can be in `base/` (for now).

Going forward, I see two reasonable options to try:

1. Move to `base/`. Do not inline. Link edir_ldapext with `base/`.
2. Move to `base/`. Inline. Assert instead of throwing. Kick the "linking" can down the road without creating a precedent.

@yadij, which option do you prefer?

Reply:
You are discarding option 3? - move to compat/xmemset_s.h, use inline and assert.

Of the two you have pre-selected. Option #1 is the closest to what I like.

Reply:
There is also option 4 - move to mem/ instead of base/

Reply:
> I had tried but failed with linkage, so many dependencies.

Thank you. I regret that your attempt confirmed my fears, but its failure resolves this change request by providing a good excuse not to inline this function -- too much out of scope work that nobody is willing to do right now.